motorboard/mbed-firmware/main.cpp

```diff
@@ -332,43 +341,68 @@ void pid() {
     }
 
     lastLoopTime = timer.read_ms();
+
+    // 2: Convert encoder values into velocity
     actualSpeedL = (metersPerTick * tickDataLeft) / dT_sec;
     actualSpeedR = (metersPerTick * tickDataRight) / dT_sec;
 
     tickDataLeft = 0;
     tickDataRight = 0;
 
+    // 3: Calculate error
     ErrorL = desiredSpeedL - actualSpeedL;
     ErrorR = desiredSpeedR - actualSpeedR;
 
-    // dPV/dt
-    dErrorL = (actual_speed_last_l - actualSpeedL)/dT_sec;
-    dErrorR = (actual_speed_last_r - actualSpeedR)/dT_sec;
+    // 4: Calculate Derivative Error
+    // TODO: Make alpha a parameter
+    constexpr float alpha = 0.75;
+    low_passed_pv_l = alpha * (actual_speed_last_l - actualSpeedL)/dT_sec + (1 - alpha) * low_passed_pv_l;
+    low_passed_pv_r = alpha * (actual_speed_last_r - actualSpeedR)/dT_sec + (1 - alpha) * low_passed_pv_r;
 
-    // sum of Error dt
+    dErrorL = low_passed_pv_l;
+    dErrorR = low_passed_pv_r;
+
+    // 5: Calculate Integral Error
+    // 5a: Calculate Error
     iErrorL += ErrorL * dT_sec;
     iErrorR += ErrorR * dT_sec;
 
-    dPWM_L = -(int)ceil((P_l * ErrorL + D_l * dErrorL + I_l * iErrorL));
-    dPWM_R = -(int)ceil((P_r * ErrorR + D_r * dErrorR + I_r * iErrorR));
+    // 5b: Perform clamping
+    // TODO: make clamping a parameter
+    constexpr float i_clamp = 255;
+    iErrorL = min(i_clamp, max(-i_clamp, iErrorL));
+    iErrorR = min(i_clamp, max(-i_clamp, iErrorR));
+
+    // 6: Sum P, I and D terms
+    float feedback_left = P_l * ErrorL + D_l * dErrorL + I_l * iErrorL;
+    float feedback_right = P_r * ErrorR + D_r * dErrorR + I_r * iErrorR;
 
-    PWM_L += dPWM_L;
-    PWM_R += dPWM_R;
+    // 7: Calculate feedforward
+    // TODO: Make Kv a parameter
+    constexpr float Kv_left = 30.511;
+    constexpr float Kv_right = 31.476;
 
-    PWM_L = min(255, max(-255, PWM_L));
-    PWM_R = min(255, max(-255, PWM_R));
+    float feedforward_left = Kv_left * desiredSpeedL;
```

Comment:
It's capped later on, so I don't think that's a problem. Also, since this is a direct map from velocity to "PWM", and the velocity commands that we are issuing are not going get close to the limits of what our motors can do, this max will actually not be reached.